staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter-config.go

```diff
@@ -0,0 +1,349 @@
+	fsi := fci.FlowSchemas()
+	reqMgr.plLister = pli.Lister()
+	reqMgr.plInformerSynced = pli.Informer().HasSynced
+	reqMgr.fsLister = fsi.Lister()
+	reqMgr.fsInformerSynced = fsi.Informer().HasSynced
+	pli.Informer().AddEventHandler(reqMgr)
+	fsi.Informer().AddEventHandler(reqMgr)
+}
+
+func (reqMgr *requestManager) triggerReload() {
+	klog.V(7).Info("triggered request-management system reloading")
+	reqMgr.configQueue.Add(0)
+}
+
+// OnAdd handles notification from an informer of object creation
+func (reqMgr *requestManager) OnAdd(obj interface{}) {
+	reqMgr.triggerReload()
+}
+
+// OnUpdate handles notification from an informer of object update
+func (reqMgr *requestManager) OnUpdate(oldObj, newObj interface{}) {
+	reqMgr.triggerReload()
+}
+
+// OnDelete handles notification from an informer of object deletion
+func (reqMgr *requestManager) OnDelete(obj interface{}) {
+	reqMgr.triggerReload()
+}
+
+func (reqMgr *requestManager) Run(stopCh <-chan struct{}) error {
+	defer reqMgr.configQueue.ShutDown()
+	klog.Info("Starting reqmgmt config controller")
+	if ok := cache.WaitForCacheSync(stopCh, reqMgr.plInformerSynced, reqMgr.fsInformerSynced); !ok {
+		return fmt.Errorf("Never achieved initial sync")
+	}
+	klog.Info("Running reqmgmt config worker")
+	wait.Until(reqMgr.runWorker, time.Second, stopCh)
+	klog.Info("Shutting down reqmgmt config worker")
+	return nil
+}
+
+func (reqMgr *requestManager) runWorker() {
+	for reqMgr.processNextWorkItem() {
+	}
+}
+
+func (reqMgr *requestManager) processNextWorkItem() bool {
+	obj, shutdown := reqMgr.configQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	func(obj interface{}) {
+		defer reqMgr.configQueue.Done(obj)
+		if !reqMgr.syncOne() {
+			reqMgr.configQueue.AddRateLimited(obj)
+		} else {
+			reqMgr.configQueue.Forget(obj)
+		}
+	}(obj)
+
+	return true
+}
+
+// syncOne attempts to sync all the config for the reqmgmt filter.  It
+// either succeeds and returns `true` or logs an error and returns
+// `false`.
+func (reqMgr *requestManager) syncOne() bool {
+	all := labels.Everything()
+	newPLs, err := reqMgr.plLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list PriorityLevelConfiguration objects: %s", err.Error())
+		return false
+	}
+	newFSs, err := reqMgr.fsLister.List(all)
+	if err != nil {
+		klog.Errorf("Unable to list FlowSchema objects: %s", err.Error())
+		return false
+	}
+	reqMgr.digestConfigObjects(newPLs, newFSs)
+	return true
+}
+
+func (reqMgr *requestManager) digestConfigObjects(newPLs []*rmtypesv1a1.PriorityLevelConfiguration, newFSs []*rmtypesv1a1.FlowSchema) {
+	oldRMState := reqMgr.curState.Load().(*requestManagerState)
+	var shareSum float64
+	newRMState := &requestManagerState{
+		priorityLevelStates: make(map[string]*priorityLevelState),
+	}
+	newlyQuiescent := make([]*priorityLevelState, 0)
+	var haveExemptPL, haveCatchAllPL, haveExemptFS, haveCatchAllFS bool
+	for _, pl := range newPLs {
+		state := oldRMState.priorityLevelStates[pl.Name]
```

Comment:
state, ok := oldRMState.priorityLevelStates[pl.Name]
if !ok { ...}

Reply:
Why make that change?  `nil` is not a value that ever gets put in that map.

Reply:
> if !ok { ...}

it looks more comforting for golang convertion :)

Reply:
@jktomer: really?  The doc says otherwise: https://golang.org/doc/faq#atomic_maps (and the first place I looked, https://golang.org/ref/mem , is pretty quiet on the issue).

Reply:
Huh, looks like I was wrong about that: m[x] does not actually add x to the map if it was not present. Feel free to disregard.